x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java

```diff
@@ -404,17 +413,90 @@ protected boolean initialRun() {
     @Override
     protected void onFinish(ActionListener<Void> listener) {
         startIndexerThreadShutdown();
-        try {
-            // This indicates an early exit since no changes were found.
-            // So, don't treat this like a checkpoint being completed, as no work was done.
-            if (hasSourceChanged == false) {
-                if (context.shouldStopAtCheckpoint()) {
-                    stop();
-                }
-                listener.onResponse(null);
-                return;
+
+        // This indicates an early exit since no changes were found.
+        // So, don't treat this like a checkpoint being completed, as no work was done.
+        if (hasSourceChanged == false) {
+            if (context.shouldStopAtCheckpoint()) {
+                stop();
             }
+            listener.onResponse(null);
+            return;
+        }
+
+        refreshDestinationIndex(new RefreshRequest(transformConfig.getDestination().getIndex()), ActionListener.wrap(response -> {
+            if (response.getFailedShards() > 0) {
+                logger.warn(
+                    "[{}] failed to refresh transform destination index, not all data might be available after checkpoint.",
+                    getJobId()
+                );
+            }
+            // delete data defined by retention policy
+            if (transformConfig.getRetentionPolicyConfig() != null) {
+                executeRetentionPolicy(listener);
+            } else {
+                finalizeCheckpoint(listener);
+            }
+        }, listener::onFailure));
+    }
 
+    private void executeRetentionPolicy(ActionListener<Void> listener) {
+        DeleteByQueryRequest deleteByQuery = RetentionPolicyToDeleteByQueryRequestConverter.buildDeleteByQueryRequest(
+            transformConfig.getRetentionPolicyConfig(),
+            transformConfig.getSettings(),
+            transformConfig.getDestination(),
+            nextCheckpoint
+        );
+
+        if (deleteByQuery == null) {
+            finalizeCheckpoint(listener);
+            return;
+        }
+
+        logger.debug(
+            new ParameterizedMessage(
```

Comment:
```suggestion
            () -> new ParameterizedMessage(
```

Nit: Use the `MessageSupplier` overload so ParameterizedMessage is not constructed at higher levels of logging